AppCenter/AppCenter/Internals/History/UserId/MSUserIdContext.m

```diff
@@ -0,0 +1,87 @@
+#import "MSAppCenterInternal.h"
+#import "MSLogger.h"
+#import "MSUserIdContextPrivate.h"
+#import "MSUtility.h"
+
+/**
+ * User Id history key.
+ */
+static NSString *const kMSUesrIdHistoryKey = @"UserIdHistory";
+
+/**
+ * Singleton.
+ */
+static MSUserIdContext *sharedInstance;
+static dispatch_once_t onceToken;
+
+@implementation MSUserIdContext
+
++ (instancetype)sharedInstance {
+  dispatch_once(&onceToken, ^{
+    if (sharedInstance == nil) {
+      sharedInstance = [[MSUserIdContext alloc] init];
+    }
+  });
+  return sharedInstance;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    NSData *data = [MS_USER_DEFAULTS objectForKey:kMSUesrIdHistoryKey];
+    if (data != nil) {
+      _userIdHistory = (NSMutableArray *)[(NSObject *)[NSKeyedUnarchiver unarchiveObjectWithData:data] mutableCopy];
+    }
+    if (!_userIdHistory) {
+      _userIdHistory = [NSMutableArray<MSUserIdHistoryInfo *> new];
+    }
+    NSUInteger count = [_userIdHistory count];
+    MSLogDebug([MSAppCenter logTag], @"%tu userId(s) in the history.", count);
+    _currentUserIdInfo = [[MSUserIdHistoryInfo alloc] initWithTimestamp:[NSDate date] andUserId:nil];
+    [_userIdHistory addObject:_currentUserIdInfo];
+  }
+  return self;
+}
+
++ (void)resetSharedInstance {
+  onceToken = 0;
+  sharedInstance = nil;
+}
+
+- (NSString *)userId {
+  return [self currentUserIdInfo].userId;
+}
+
+- (void)setUserId:(nullable NSString *)userId {
+  @synchronized(self) {
+    [self.userIdHistory removeLastObject];
```

Comment:
There's still no comment here - please clarify why we are removing history.

Reply:
I thought you want to add it in the PR. 😄 
Added that.